var/spack/repos/builtin/packages/gcc/package.py

```diff
@@ -254,6 +256,63 @@ class Gcc(AutotoolsPackage, GNUMirrorPackage):
 
     build_directory = 'spack-build'
 
+    @property
+    def executables(self):
+        names = [r'gcc', r'[^\w]?g\+\+', r'gfortran']
+        suffixes = [r'-mp-\d\.\d', r'-\d\.\d', r'-\d', r'\d\d']
+        return [r''.join(x) for x in itertools.product(names, suffixes)]
+
+    @classmethod
+    def determine_spec_details(cls, prefix, exes_in_prefix):
+        exes_by_version = collections.defaultdict(list)
+        for exe in exes_in_prefix:
+            # clang++ matches g++ -> clan[g++]
+            if 'clang' in exe:
+                continue
+
+            version_str = cls.determine_version(exe)
+            if version_str:
+                exes_by_version[version_str].append(exe)
+
+        specs = []
+        for version_str, exes in exes_by_version.items():
+            spec_str = cls.determine_variants(exes, version_str)
+            if spec_str:
+                specs.append(Spec(spec_str))
+
+        return specs
```

Comment:
@scheibelp @tgamblin Would a reworking of this method be a good default implementation for `determine_spec_details` in your opinion?  Modulo the `clang` check, which might be turned into a generic `filter_exes` step, I believe many packages will need to do the same thing as here for detection. 

I am wondering if we could consider `determine_version` and optionally `determine_variants` as simpler methods to require user to implement, and leave `determine_spec_details` as the big hammer to be only used when needed. Thoughts?

Reply:
> Would a reworking of this method be a good default implementation for determine_spec_details in your opinion? Modulo the clang check, which might be turned into a generic filter_exes step, I believe many packages will need to do the same thing as here for detection.

I agree it looks like this could be simplified for many packages. I'm not sure what `filter_exes` would look like, but I agree that the `clang` check could be generalized (especially when a package only defines a single executable).